src/plugin/PluginManager.php

```diff
@@ -653,8 +677,25 @@ public function registerEvent(string $event, \Closure $handler, int $priority, P
 
 		$timings = Timings::getEventHandlerTimings($event, $handlerName, $plugin->getDescription()->getFullName());
 
-		$registeredListener = new RegisteredListener($handler, $priority, $plugin, $handleCancelled, $timings);
+		if(is_subclass_of($event, AsyncEvent::class) && $this->canHandleAsyncEvent($handler)){
+			$registeredListener = new RegisteredAsyncListener($handler, $priority, $plugin, $handleCancelled, $noConcurrentCall, $timings);
+		}else{
+			$registeredListener = new RegisteredListener($handler, $priority, $plugin, $handleCancelled, $timings);
+		}
 		HandlerListManager::global()->getListFor($event)->register($registeredListener);
 		return $registeredListener;
 	}
+
+	/**
+	 * Check if the given handler return type is async-compatible (equal to Promise)
+	 *
+	 * @phpstan-template TEvent of Event
+	 * @phpstan-param (\Closure(TEvent) : void)|(\Closure(AsyncEvent&TEvent) : Promise<null>) $handler
+	 */
+	private function canHandleAsyncEvent(\Closure $handler) : bool {
+		$reflection = new \ReflectionFunction($handler);
+		$return = $reflection->getReturnType();
+
+		return $return instanceof \ReflectionNamedType && $return->getName() === Promise::class;
```

Comment:
Any thoughts on making async handlers opt-in in general, and just check the return type to double-confirm? i.e. either handler return type is `Promise`, it does not return a `Promise` object and it does not declare `@async`, or it declares `@async` and return type is `Promise`; if it has a return type hint of `Promise` or the function invocation returns a `Promise`, the server should crash.

Reply:
As async events are now completely detached from synchronous events, there is no longer any need for double confirmation, as each event handler will behave in the same way.